pkg/sources/kafka/reader.go

```diff
@@ -445,37 +485,30 @@ func (r *KafkaSource) startConsumer() {
 }
 
 func toReadMessage(m *sarama.ConsumerMessage) *isb.ReadMessage {
-	offset := toOffset(m.Topic, m.Partition, m.Offset)
+	readOffset := &kafkaOffset{
+		offset:       m.Offset,
+		partitionIdx: m.Partition,
+		topic:        m.Topic,
+	}
 	msg := isb.Message{
 		Header: isb.Header{
 			MessageInfo: isb.MessageInfo{EventTime: m.Timestamp},
-			ID:          offset,
+			ID:          readOffset.String(),
 			Keys:        []string{string(m.Key)},
 		},
 		Body: isb.Body{Payload: m.Value},
 	}
 
 	return &isb.ReadMessage{
-		ReadOffset: isb.SimpleStringOffset(func() string { return offset }),
+		ReadOffset: readOffset,
 		Message:    msg,
 	}
 }
 
-func toOffset(topic string, partition int32, offset int64) string {
-	// TODO handle this elegantly
-	return fmt.Sprintf("%s:%v:%v", topic, partition, offset)
-}
-
-func offsetFrom(offset string) (string, int32, int64, error) {
-	tokens := strings.Split(offset, ":")
-	if len(tokens) < 3 {
-		return "", 0, 0, errors.New("malformed offset")
-	}
-	var poffset, partition int64
-	var err error
-	poffset, err = strconv.ParseInt(tokens[2], 10, 64)
-	if err == nil {
-		partition, err = strconv.ParseInt(tokens[1], 10, 32)
+func extractTopicFromOffset(offset isb.Offset) (string, error) {
```

Comment:
can't we just get it from kafka.Offset? just by type casting? why do we need string manipulation?

Reply:
delete this func then `extractTopicFromOffset`?